### ISSUE TITLE
Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105

### ISSUE THREAD

**Issue by GAgafencu** [statOmics/tradeSeq#120]
Dear tradeSeq team,

Thanks for providing this interesting analysis tool for the community. I am trying to establish to use this tool to understand gene-expression patterns across and and between trajectories. After running this piece of code on a SingleCellExperiment object generated from a Seurat one that contains scRNA-seq data from 74430 cells generated using a 10XGenomics Chromium V2 kit:

sce <- as.SingleCellExperiment(samples_merged_use, assay = "RNA")
sce <- slingshot(sce, reducedDim = 'UMAP', clusterLabels = sce@colData$seurat_clusters, start.clus = '14')
sce <- fitGAM(counts = assays(sce)$counts, sds = SlingshotDataSet(sce), conditions = as.factor(colData(sce)$age_group), nknots = 6)

,I get this error:

Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'counts' in selecting a method for function 'fitGAM': Cholmod error 'problem too large' at file ../Core/cholmod_dense.c, line 105

It seems like task can't be handled due to memory capacity issues, but I'm running this on a computing node that has 40 CPU cores and 1TB of RAM. Could you please let me know how this issue can be fixed?

Thank you for your help!

**Comment by koenvandenberge** [comment]
Thank you for reporting @GAgafencu. 

Is the count matrix in the `SingleCellExperiment` object a sparse count matrix? 
Does the error persist if you're using a lower number of genes for the fitting? You can, for example, try fitting the first 100 genes by setting the `genes=1:100` argument in `fitGAM`.

That would help us diagnose the problem.
Thanks,
Koen

**Comment by GAgafencu** [comment]
Thanks Koen for the quick reply!

I've checked the nature of the count matrix in the SingleCellExperiment object using  class(assays(sce)$counts)
and I got this output:
[1] "dgCMatrix"
attr(,"package")
[1] "Matrix"
It seems to be a sparse matrix to me.

I get the same error when I run the fitGAM command with the argument genes=1:100.

Would it make any sense to run fitGAM on a subset of the data (cell randomly picked from the original sce object, half of the size of the original sce object in terms of cell numbers) and use that model to go further in my analysis? It might not make sense since every cell in the original sce object must be fitted to a GAM. I would wish to fit the GAM on the whole dataset not a subset.

Thank you for your help and I am looking forward to your reply.

This is the output of the sessionInfo() command
R version 4.0.1 (2020-06-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /Filers/package/R-base/4.0.1/lib64/R/lib/libRblas.so
LAPACK: /Filers/package/R-base/4.0.1/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] tradeSeq_1.5.06             slingshot_1.6.1             princurve_2.1.4             scran_1.16.0               
 [5] SingleCellExperiment_1.10.1 SingleR_1.2.4               SummarizedExperiment_1.18.2 DelayedArray_0.14.1        
 [9] matrixStats_0.56.0          Biobase_2.48.0              GenomicRanges_1.40.0        GenomeInfoDb_1.24.2        
[13] IRanges_2.22.2              S4Vectors_0.26.1            BiocGenerics_0.34.0         harmony_1.0                
[17] Rcpp_1.0.5                  bmcite.SeuratData_0.3.0     SeuratData_0.2.1            SeuratWrappers_0.3.0       
[21] sctransform_0.3.1           Seurat_3.9.9.9010           corrplot_0.84               pheatmap_1.0.12            
[25] scales_1.1.1                viridis_0.5.1               viridisLite_0.3.0           plotly_4.9.2.1             
[29] reshape_0.8.8               RColorBrewer_1.1-2          forcats_0.5.0               stringr_1.4.0              
[33] dplyr_1.0.2                 purrr_0.3.4                 readr_1.3.1                 tidyr_1.1.2                
[37] tibble_3.0.3                ggplot2_3.3.2               tidyverse_1.3.0            

loaded via a namespace (and not attached):
  [1] reticulate_1.16               tidyselect_1.1.0              RSQLite_2.2.0                
  [4] AnnotationDbi_1.50.1          htmlwidgets_1.5.1             docopt_0.6.1                 
  [7] combinat_0.0-8                grid_4.0.1                    BiocParallel_1.22.0          
 [10] Rtsne_0.15                    munsell_0.5.0                 codetools_0.2-16             
 [13] ica_1.0-2                     statmod_1.4.34                future_1.20.1                
 [16] miniUI_0.1.1.1                withr_2.2.0                   fastICA_1.2-2                
 [19] colorspace_1.4-1              rstudioapi_0.11               ROCR_1.0-11                  
 [22] tensor_1.5                    listenv_0.8.0                 slam_0.1-47                  
 [25] GenomeInfoDbData_1.2.3        polyclip_1.10-0               bit64_0.9-7                  
 [28] parallelly_1.21.0             vctrs_0.3.4                   generics_0.0.2               
 [31] BiocFileCache_1.12.0          R6_2.5.0                      ggbeeswarm_0.6.0             
 [34] rsvd_1.0.3                    VGAM_1.1-3                    locfit_1.5-9.4               
 [37] bitops_1.0-6                  spatstat.utils_1.17-0         assertthat_0.2.1             
 [40] promises_1.1.1                beeswarm_0.2.3                gtable_0.3.0                 
 [43] globals_0.13.1                goftest_1.2-2                 rlang_0.4.7                  
 [46] splines_4.0.1                 lazyeval_0.2.2                broom_0.7.0                  
 [49] BiocManager_1.30.10           yaml_2.2.1                    reshape2_1.4.4               
 [52] abind_1.4-5                   modelr_0.1.8                  backports_1.1.9              
 [55] httpuv_1.5.4                  tools_4.0.1                   ellipsis_0.3.1               
 [58] ggridges_0.5.2                plyr_1.8.6                    zlibbioc_1.34.0              
 [61] densityClust_0.3              RCurl_1.98-1.2                rpart_4.1-15                 
 [64] deldir_0.1-25                 pbapply_1.4-3                 cowplot_1.1.0                
 [67] zoo_1.8-8                     haven_2.3.1                   ggrepel_0.8.2                
 [70] cluster_2.1.0                 fs_1.5.0                      magrittr_2.0.1               
 [73] data.table_1.13.0             lmtest_0.9-37                 reprex_0.3.0                 
 [76] RANN_2.6.1                    fitdistrplus_1.1-3            hms_0.5.3                    
 [79] patchwork_1.0.0               mime_0.9                      xtable_1.8-4                 
 [82] sparsesvd_0.2                 readxl_1.3.1                  gridExtra_2.3                
 [85] HSMMSingleCell_1.8.0          compiler_4.0.1                scater_1.16.2                
 [88] KernSmooth_2.23-17            crayon_1.3.4                  htmltools_0.5.0              
 [91] mgcv_1.8-31                   later_1.1.0.1                 lubridate_1.7.9              
 [94] DBI_1.1.0                     ExperimentHub_1.14.2          dbplyr_1.4.4                 
 [97] MASS_7.3-51.6                 rappdirs_0.3.1                Matrix_1.2-18                
[100] cli_2.0.2                     igraph_1.2.5                  pkgconfig_2.0.3              
[103] xml2_1.3.2                    vipor_0.4.5                   dqrng_0.2.1                  
[106] XVector_0.28.0                rvest_0.3.6                   digest_0.6.25                
[109] RcppAnnoy_0.0.18              DDRTree_0.1.5                 spatstat.data_1.4-3          
[112] cellranger_1.1.0              leiden_0.3.3                  uwot_0.1.9.9000              
[115] edgeR_3.30.3                  DelayedMatrixStats_1.10.1     curl_4.3                     
[118] shiny_1.5.0                   monocle_2.16.0                lifecycle_0.2.0              
[121] nlme_3.1-148                  jsonlite_1.7.0                BiocNeighbors_1.6.0          
[124] limma_3.44.3                  fansi_0.4.1                   pillar_1.4.6                 
[127] lattice_0.20-41               fastmap_1.0.1                 httr_1.4.2                   
[130] survival_3.1-12               interactiveDisplayBase_1.26.3 glue_1.4.2                   
[133] remotes_2.1.1                 qlcMatrix_0.9.7               FNN_1.1.3                    
[136] spatstat_1.64-1               png_0.1-7                     BiocVersion_3.11.1           
[139] bit_4.0.4                     stringi_1.4.6                 blob_1.2.1                   
[142] BiocSingular_1.4.0            AnnotationHub_2.20.0          memoise_1.1.0                
[145] ape_5.4                       irlba_2.3.3                   future.apply_1.5.0

**Comment by koenvandenberge** [comment]
Thank you @GAgafencu.
Do you mind sharing the subset of your data corresponding to e.g. the first 100 genes and all relevant input to `fitGAM`, so we can look into the issue?
Feel free to share e.g. a Dropbox folder or send an e-mail to koenvdberge@berkeley.edu.

**Comment by koenvandenberge** [comment]
Hi @GAgafencu 

Are you encountering this error immediately?
I just tried this locally and it’s running and has fitted for a few genes.
It’s very slow though, since you’re estimating 8 lineages * 2 conditions * (default of) 6 knots = 96 parameters for each gene, especially since the parameters will be very correlated. You may want to consider tuning this by e.g. reconsidering the number of lineages / knots.

**Comment by GAgafencu** [comment]
The error occurs immediately after I initiate the command when running on the whole dataset even with genes=1:100 on a a computing node that has 40 CPU cores and 1TB of RAM .

I've just split my dataset in half by randomly selected approx. 35k cells and set genes = 1:300 with a k=6 and the fitGAM function started to execute and it estimated around 1.5 days until it completes the task. Suffice to say I stopped its execution and now I will reconsider the number of lineages that I want to investigate as not all the lineages identified by Slingshot make sense from a biological standpoint and I am running the evaluateK() function on the subsetted data just to get a feeling of an optimal k. 
Ideally, I would like to process the entire dataset as a whole, but let me know if there is away to solve the issue that I raised.
Thank you for your help!

**Comment by koenvandenberge** [comment]
Thanks. It seems that the problem is not with `tradeSeq` then, but rather with the server and input you're providing to `fitGAM`. 
We definitely do not recommend fitting this many lineages if some don't make sense biologically as it could have a detrimental effect on the downstream results.
I will close this issue for now, feel free to reopen if other issues arise.

**Comment by GAgafencu** [comment]
Thank you for your due diligence in addressing this issue. If it is not too much to ask, could you please elaborate on if by the input to this function is to be one of the sources of this issue are you solely referring to the fact that I should focus only on the relevant lineages or there is something with the count matrix that is causing this error? If it is something with the count matrix what would be the property of the matrix that is generating this issue?
Thank you for your help!

**Comment by koenvandenberge** [comment]
Hi @GAgafencu 

Since the function runs when you subset your data, it seemed to me like the problem was not with `tradeSeq`, but possibly with e.g. how the server is set up or how you submit your job.
When you subsetted your data, did you also only provide the `pseudotime` and `cellWeights` of those subsetted cells?

Note also that since you don't seem to be taking full advantage of the heavy computer you're using. You'd have to allow parallelization, see this part in our vignette: https://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/fitGAM.html#parallel-computing.
For debugging I don't recommend parallelization, though, but that would allow you to take advantage of the many cores.

I'm not able to reproduce the issue based on what you've shared, so it's hard for me to solve your issue.

**Comment by GAgafencu** [comment]
The error I reported definitely has something to do with how the computing node is configured as this function is stops with a different error if I try to use the parallelisation option that you mention. When I was running the function on a subset of cells, I was providing both the pseudotime and cellWeights values for that cell subset as well as the normalised counts for those cells.
I eventually got the function to work, even though I did not waited for it to finish, by running it on the whole dataset, but only on a couple of lineages and using only a subset of genes.  As you mentioned, my error has more to do with the computing infrastructure than the package and I'll keep you updated in case it crops out again in a different setting.
